calculator_test.go

```diff
@@ -2,23 +2,205 @@ package calculator_test
-		t.Errorf("want %f, got %f", want, got)
+	testCases := []struct {
+		a, b, want float64
+		nums       []float64
+		name       string
+	}{
+		{a: 5, b: 6, nums: []float64{}, want: 11, name: "5 + 6 = 11. Empty nums."},
+		{a: 1, b: 2, nums: []float64{3}, want: 6, name: "1 + 2 + 3 = 6"},
+		{a: 1, b: 2, nums: []float64{3, 4}, want: 10, name: "1 + 2 + 3 + 4 = 10"},
+	}
+
+	for _, tc := range testCases {
+		got := calculator.Add(tc.a, tc.b, tc.nums...)
+
+		if tc.want != got {
+			t.Errorf("%v - want %.1f, got %.1f", tc.name, tc.want, got)
+		}
+	}
+}
+
+func TestAddRandom(t *testing.T) {
+	t.Parallel()
+	for i := 0; i < 10000000; i++ {
+		a := rand.Float64()
+		b := rand.Float64()
+		want := a + b
+		got := calculator.Add(a, b)
+
+		if want != got {
+			t.Errorf("%g + %g - want: %.2f, got %.2f", a, b, want, got)
+		}
+	}
+}
+
+func TestSubtract(t *testing.T) {
+	t.Parallel()
+
+	testCases := []struct {
+		a, b, want float64
+		nums       []float64
+		name       string
+	}{
+		{a: 5, b: 5, nums: []float64{}, want: 0, name: "5 - 5 = 0. Empty nums."},
+		{a: 3, b: 2, nums: []float64{1}, want: 0, name: "3 - 2 - 1 = 0"},
+		{a: 100, b: 8, nums: []float64{1, 12}, want: 79, name: "100 - 8 - 12 = 79"},
+		{a: 1, b: 8, nums: []float64{10}, want: -17, name: "1 - 8 - 10 = -17"},
+	}
+
+	for _, tc := range testCases {
+		got := calculator.Subtract(tc.a, tc.b, tc.nums...)
+
+		if tc.want != got {
+			t.Errorf("%v - want %.1f, got %.1f", tc.name, tc.want, got)
+		}
 	}
 }
 
-// func TestSubtract(t *testing.T) {
-// 	t.Parallel()
-// 	var want float64 = 2
-// 	got := calculator.Subtract(4, 2)
-// 	if want != got {
-// 		t.Errorf("want %f, got %f", want, got)
-// 	}
-// }
+func TestMultiply(t *testing.T) {
+	t.Parallel()
+
+	testCases := []struct {
+		a, b, want float64
+		nums       []float64
+		name       string
+	}{
+		{a: 4, b: 11, nums: []float64{}, want: 44, name: "4 * 11 = 44. Empty nums."},
+		{a: 100, b: 8, nums: []float64{12}, want: 9600, name: "100 * 8 * 12 = 9600"},
+		{a: 3, b: 2, nums: []float64{1, -10}, want: -60, name: "3 * 2 * -10 = -60"},
+		{a: 0, b: 10, nums: []float64{55, 10, 2}, want: 0, name: "0 * 10 * 55 * 10 * 2 = 0. 0 passed as first parameter."},
+		{a: 400, b: 13, nums: []float64{0, 4}, want: 0, name: "400 * 13 * 0 * 4 = 0"},
+	}
+
+	for _, tc := range testCases {
+		got := calculator.Multiply(tc.a, tc.b, tc.nums...)
+
+		if tc.want != got {
+			t.Errorf("%v - want %.1f, got %.1f", tc.name, tc.want, got)
+		}
+	}
+
+}
+
+func TestDivide(t *testing.T) {
+	t.Parallel()
+
+	testCases := []struct {
+		a, b, want  float64
+		nums        []float64
+		name        string
+		errExpected bool
+	}{
+		{a: 7, b: 0, nums: []float64{10, 44}, want: 0, name: "1 / 8 / -10 = -0.0125", errExpected: true},
+		{a: 3, b: 2, nums: []float64{}, want: 1.5, name: "3 / 2 = 1.5", errExpected: false},
+		{a: 2, b: 3, nums: []float64{}, want: 0.6666666666666666, name: "Rational value that doesn't have an exact floating-point representation", errExpected: false},
```

Comment:
But this isn't correct, is it? Two divided by three is _not_ 0.6666666666666666, it's 2/3.

Reply:
Ok - so, I have addressed this in 9e033a3 and  671df14, but I am not 100% satisfied with the error messages. 
An example would be passing: 
```
		{a: 1, b: 3, roundedResult: 0.6666, tolerance: 0.001},
```
as a test struct to it, and I would get an error that says `Result(0.6666) outside tolerance range(0.001000000)` which, whilst technically correct, is not very informative.

Am I making my life more complicated than it should be here?

Reply:
I would just say something like 'result is not close enough'.

Reply:
All right - that makes sense, fixed it a00118c :)